crates/sui/src/upgrade_compatibility.rs

```diff
@@ -0,0 +1,414 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Error};
+use std::collections::BTreeSet;
+use thiserror::Error;
+
+use move_binary_format::{
+    compatibility::Compatibility,
+    compatibility_mode::CompatibilityMode,
+    file_format::AbilitySet,
+    normalized::{Enum, Function, Module, Struct},
+    CompiledModule,
+};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::{IdentStr, Identifier},
+    language_storage::ModuleId,
+};
+use sui_json_rpc_types::{SuiObjectDataOptions, SuiRawData};
+use sui_protocol_config::ProtocolConfig;
+use sui_sdk::SuiClient;
+use sui_types::{base_types::ObjectID, execution_config_utils::to_binary_config};
+
+/// Check the upgrade compatibility of a new package with an existing on-chain package.
+pub async fn check_compatibility(
+    client: &SuiClient,
+    package_id: ObjectID,
+    compiled_modules: &[Vec<u8>],
+    protocol_config: ProtocolConfig,
+) -> Result<(), Error> {
+    let new_modules = compiled_modules
+        .iter()
+        .map(|b| CompiledModule::deserialize_with_config(b, &to_binary_config(&protocol_config)))
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(|_| anyhow!("Unable to check compatibility of new module"))?;
+
+    let existing_obj_read = client
+        .read_api()
+        .get_object_with_options(package_id, SuiObjectDataOptions::new().with_bcs())
+        .await
+        .map_err(|_| anyhow!("Unable to get existing module"))?;
+
+    let existing_obj = existing_obj_read
+        .into_object()
+        .map_err(|_| anyhow!("Unable to read object"))?
+        .bcs
+        .ok_or_else(|| anyhow!("Unable to read object"))?;
+
+    let existing_package = match existing_obj {
+        SuiRawData::Package(pkg) => Ok(pkg),
+        SuiRawData::MoveObject(_) => Err(anyhow!("Object found when package expected")),
+    }?;
+
+    let existing_modules = existing_package
+        .module_map
+        .iter()
+        .map(|m| CompiledModule::deserialize_with_config(m.1, &to_binary_config(&protocol_config)))
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(|_| anyhow!("Unable to check compatibility of existing module"))?;
+
+    // for each existing find the new one run compatibility check
+    for existing_module in existing_modules {
+        let name = existing_module.self_id().name().to_owned();
+        // find the new module with the same name
+        match new_modules.iter().find(|m| m.self_id().name().eq(&name)) {
+            Some(new_module) => {
+                (Compatibility {
```

Comment:
nit: Parens should be unnecessary here

Reply:
Also, it would be good to make that change we talked about of having a constructor to build a `Compatibility` struct with all the right flags.